### ISSUE TITLE
getaddrinfo result ordering wrong with ares resolver

### ISSUE THREAD

**Issue by icgood** [gevent/gevent#815]
* gevent version: 1.1.1
* Python version: 2.7.9
* Operating System: Debian 8.4

### Description:

The results of `gevent.socket.getaddrinfo` are in the wrong order, when using the `ares` DNS resolver. They show IPv4 results first, when the thread resolver shows IPv6 results first.

### What I've run:

Ares resolver:
```
$ GEVENT_RESOLVER=ares python
Python 2.7.9 (default, Mar  1 2015, 12:57:24)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from gevent import socket
>>> socket.getaddrinfo('gmail-smtp-in.l.google.com', 25)
[(2, 1, 6, '', ('209.85.144.27', 25)), (2, 2, 17, '', ('209.85.144.27', 25)), (2, 3, 0, '', ('209.85.144.27', 25)), (10, 1, 6, '', ('2607:f8b0:400d:c0e::1a', 25, 0, 0)), (10, 2, 17, '', ('2607:f8b0:400d:c0e::1a', 25, 0, 0)), (10, 3, 0, '', ('2607:f8b0:400d:c0e::1a', 25, 0, 0))]
```

Thread resolver:
```
$ python
Python 2.7.9 (default, Mar  1 2015, 12:57:24)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from gevent import socket
>>> socket.getaddrinfo('gmail-smtp-in.l.google.com', 25)
[(10, 1, 6, '', ('2607:f8b0:400d:c0e::1b', 25, 0, 0)), (10, 2, 17, '', ('2607:f8b0:400d:c0e::1b', 25, 0, 0)), (10, 3, 0, '', ('2607:f8b0:400d:c0e::1b', 25, 0, 0)), (2, 1, 6, '', ('209.85.144.26', 25)), (2, 2, 17, '', ('209.85.144.26', 25)), (2, 3, 0, '', ('209.85.144.26', 25))]
```

**Comment by jamadden** [comment]
I don't think the order of results from `getaddrinfo` is guaranteed. For example, on one of my systems, plain python returns IPV4 first:
```
$ python
Python 2.7.10 (246c9cf22037, Mar 09 2016, 16:14:57)
[PyPy 5.0.0 with GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>> import socket
>>>> socket.getaddrinfo('gmail-smtp-in.l.google.com', 25)
[(2, 2, 17, '', ('64.233.168.26', 25)), (2, 1, 6, '', ('64.233.168.26', 25)), (30, 2, 17, '', ('2607:f8b0:4001:c1c::1b', 25, 0L, 0L)), (30, 1, 6, '', ('2607:f8b0:4001:c1c::1b', 25, 0L, 0L))]
```
```
python3.5
Python 3.5.1 (default, Mar  2 2016, 03:38:02)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.getaddrinfo('gmail-smtp-in.l.google.com', 25)
[(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('64.233.168.26', 25)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('64.233.168.26', 25)), (<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('2607:f8b0:4001:c1c::1b', 25, 0, 0)), (<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2607:f8b0:4001:c1c::1b', 25, 0, 0))]
```

I do get IPV6 first on a different system.

The [linux man page](http://man7.org/linux/man-pages/man3/getaddrinfo.3.html) references a sorting algorithm defined by [RFC 3484](https://www.ietf.org/rfc/rfc3484.txt), but that RFC notes that it can be customized and overridden and otherwise changed on a system-by-system basis.

**Comment by icgood** [comment]
I believe there is a higher precedence RFC that, though I am generalizing greatly here, IPv6 be listed before IPv4:

https://tools.ietf.org/html/rfc6724#section-10.3

**Comment by jamadden** [comment]
Be that as it may, it's certainly not universally respected. Apple platforms make one obvious example. 

I haven't absolutely confirmed, but I think gevent doesn't have any control over that, leaving it up to c-ares.

**Comment by icgood** [comment]
@jamadden The order is controlled by this line in gevent:

https://github.com/gevent/gevent/blob/master/src/gevent/resolver_ares.py#L214

In an ideal world, c-ares _would_ implement `getaddrinfo` and it seems like [they plan to](http://c-ares.haxx.se/why.html). I definitely don't mean to suggest that gevent should re-implement all the logic a system uses to decide the order as per RFC-6724, but if it were up to me returning v6 first is one step I'd take. ¯\\\_(ツ)\_/¯ 

Feel free to close.

**Comment by jamadden** [comment]
You are right that gevent currently does control the returned resolution order, I spoke before I confirmed.

In addition to OS X (v4 first) and linux (v6) first, I've checked the system Python on FreeBSD 10 (v4 first) and Solaris 11 (v6 first). As half do it one way and half the other, it appears that there is no strong consensus yet.

For the time being, for the sake of BWC, I will keep things as they are (but I'll add a note and maybe some docs). Maybe it can be revisited in the (medium) future, if c-ares doesn't provide a stronger implementation.

Thank you for reporting this!
